renderdoc/driver/gl/gl_pixelhistory.cpp

```diff
@@ -0,0 +1,225 @@
+
+  // Allocate a texture for the pixel history depth/stencil values
+  driver->glGenTextures(1, &resources.dsImage);
+  driver->glBindTexture(eGL_TEXTURE_2D, resources.dsImage);
+  driver->CreateTextureImage(resources.dsImage, eGL_DEPTH32F_STENCIL8, eGL_NONE, eGL_NONE,
+                             eGL_TEXTURE_2D, 2, desc.width >> sub.mip, desc.height >> sub.mip, 1,
+                             desc.msSamp, 1);
+  driver->glFramebufferTexture(eGL_FRAMEBUFFER, eGL_DEPTH_STENCIL_ATTACHMENT, resources.dsImage, 0);
+
+  driver->glGenQueries(1, &resources.occlusionQuery);
+
+  return true;
+}
+
+bool PixelHistoryDestroyResources(WrappedOpenGL *driver, const GLPixelHistoryResources &resources)
+{
+  driver->glDeleteTextures(1, &resources.colorImage);
+  driver->glDeleteTextures(1, &resources.dsImage);
+  driver->glDeleteFramebuffers(1, &resources.frameBuffer);
+
+  driver->glDeleteQueries(1, &resources.occlusionQuery);
+
+  return true;
+}
+
+rdcarray<EventUsage> QueryModifyingEvents(WrappedOpenGL *driver, GLPixelHistoryResources &resources,
+                                          const rdcarray<EventUsage> &events, int x, int y)
+{
+  driver->glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT | GL_STENCIL_BUFFER_BIT);
+  driver->ReplayLog(0, events[0].eventId, eReplay_WithoutDraw);
+
+  rdcarray<EventUsage> modEvents;
+
+  for(size_t i = 0; i < events.size(); i++)
+  {
+    if(events[i].usage == ResourceUsage::Clear || isDirectWrite(events[i].usage))
+    {
+      modEvents.push_back(events[i]);
+    }
+    else
+    {
+      driver->glBindFramebuffer(eGL_FRAMEBUFFER, resources.frameBuffer);
+
+      driver->glDisable(eGL_DEPTH_TEST);
+      driver->glDisable(eGL_STENCIL_TEST);
+      driver->glEnable(eGL_SCISSOR_TEST);
+      driver->glScissor(x, y, 1, 1);
+
+      driver->glBeginQuery(eGL_ANY_SAMPLES_PASSED, resources.occlusionQuery);
+      driver->ReplayLog(events[i].eventId, events[i].eventId, eReplay_OnlyDraw);
+
+      int result;
+      driver->glGetQueryObjectiv(resources.occlusionQuery, eGL_QUERY_RESULT, &result);
+      if(result)
+      {
+        modEvents.push_back(events[i]);
+      }
+    }
+
+    if(i < events.size() - 1)
+    {
+      driver->ReplayLog(events[i].eventId + 1, events[i + 1].eventId, eReplay_WithoutDraw);
+    }
+  }
+  return modEvents;
+}
+
+void QueryPostModPixelValues(WrappedOpenGL *driver, GLPixelHistoryResources &resources,
+                             const rdcarray<EventUsage> &modEvents, int x, int y,
+                             rdcarray<PixelModification> &history)
+{
+  driver->glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT | GL_STENCIL_BUFFER_BIT);
+  driver->ReplayLog(0, modEvents[0].eventId, eReplay_WithoutDraw);
+
+  for(size_t i = 0; i < modEvents.size(); i++)
+  {
+    driver->glBindFramebuffer(eGL_FRAMEBUFFER, resources.frameBuffer);
+    driver->ReplayLog(modEvents[i].eventId, modEvents[i].eventId, eReplay_Full);
```

Comment:
Since this is GL you should restore the framebuffer binding after replaying the draw, since e.g. one of the events below between draws might interact with the current framebuffer.